ledger/test-common/src/main/daml/semantic/ValueNesting.daml

```diff
@@ -0,0 +1,89 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module ValueNesting where
+
+data Nat = Z | S Nat
+  deriving (Eq, Show)
+
+construct: Int -> Nat -> Nat
+construct x acc | x <= 0 = acc
+construct x acc  = construct (x-1) (S acc)
+
+toNat : Int -> Nat
+toNat x = construct x Z
+
+destruct: Nat -> Int -> Int
+destruct Z acc = acc
+destruct (S n) acc = destruct n (acc + 1)
+
+toInt: Nat -> Int
+toInt x = destruct x 0
+
+template Contract
+  with
+    party: Party
+    i: Int
+    n: Nat
+  where
+    signatory party
+
+template ContractWithKey
+  with
+    party: Party
+    i: Int
+  where
+    signatory party
+    key (party, toNat i): (Party, Nat)
+    maintainer key._1
+
+template Handler
+  with
+    party : Party
+  where
+    signatory party
+    controller party can
+      nonconsuming Construct : Nat
+        with i : Int
+        do
+          pure $ toNat i
+      nonconsuming Destruct : Int
+        with n: Nat
+        do pure $ toInt n
+      nonconsuming ContructThenDestruct : Int
+        with
+          i: Int
+        do
+          n <- exercise self Construct with i = i
+          exercise self Destruct with n = n
+      nonconsuming Create: ContractId Contract
```

Comment:
Don't we also need a test case where the exercise argument itself is a `Nat`?

Reply:
That’s `Destruct` in line 50

Reply:
`ConstructThenDestruct` test the case where the exercise argument itself is a `Nat`.

Note that in my original design all the `Nat` were built by evaluation and not coming for the ledger API. 
This is because, I am mainly concerned  by the depth of value in the output transaction and I trust the check of value depth in the command is working properly.

Though, I added in e58d392c7f one test for create command and one for exercise command that contain deeply nested value as it is pretty straightforward.

Reply:
Add 2 tests for `CreateAndExercise`